apps/omg/lib/omg/state/core.ex

```diff
@@ -388,6 +432,49 @@ defmodule OMG.State.Core do
     }
   end
 
+  defp collect_fees(%Core{fees_paid: fees_paid} = state, token_surplussed) do
+    fees_paid_with_new =
+      token_surplussed
+      |> Enum.reject(fn {_token, amount} -> amount == 0 end)
+      |> Map.new()
+
+    %Core{
+      state
+      | fees_paid:
+          Map.merge(fees_paid, fees_paid_with_new, fn
+            _token, collected, tx_surplus -> collected + tx_surplus
+          end)
+    }
+  end
+
+  defp disallow_payments(state), do: %Core{state | fee_claiming_started: true}
+
+  defp flush_collected_fees_for_token(state, %Output{currency: token}),
+    do: %Core{state | fees_paid: Map.delete(state.fees_paid, token)}
+
+  @spec claim_fees(state :: t()) :: t()
+  defp claim_fees(
```

Comment:
um, wait a second, now after the cleanups I'm seeing sth strange here. Why aren't we calling `exec` from `OMG.State.form_block`? It looks like if the Watcher calls into here it will `exec` the fee transactions twice (or do sth comparably strange).

Reply:
This is very good catch sir :exclamation: